Add pool DB operation `removeRetiredPools`.

jonathanknowles opened the pull request:
# Issue Number

#2018 

# Overview

Building on the work added in PRs #2024 and #2038, this PR:

- [x] Adds a `removeRetiredPools` operation to the pool DB layer:<br><br>
    ```hs
    removeRetiredPools  
        :: EpochNo
        -> stm [PoolRetirementCertificate]
        -- ^ Remove all pools with an active retirement epoch that is earlier 
        -- than or equal to the specified epoch.  
        --
        -- Returns the retirement certificates of the pools that were removed.
    ```
- [x] Writes a bracketed `MsgRemovingRetiredPools` message to the log whenever the `removeRetiredPools` operation is called:<br><br>
    ```
    Looking for pools that retired in or before epoch 1000.
    Removing the following retired pools:
        Pool 4355a46b with retirement epoch 999
        Pool 53c234e5 with retirement epoch 999
        Pool 121cfccd with retirement epoch 1000
        Pool 917df332 with retirement epoch 1000
    Finished removing retired pools.
    ```

# Further Work

Before garbage collection can actually happen, the wallet needs to call `removeRetiredPools` at appropriate intervals. This work is deferred to a future PR.

See issue #2019.

jonathanknowles committed:
Provide `removeRetiredPools` DB operation.

jonathanknowles committed:
Add `MsgGarbageCollection` logging constructor.

jonathanknowles committed:
Add logging to `removeRetiredPools`.

jonathanknowles committed:
Revise `Buildable` instance for `PoolRetirementCertificate`.

Grammatically, the generated text now makes sense regardless of whether
the retirement is in the future or the past.

jonathanknowles committed:
Rename `retiredIn` to `retirementEpoch`.

Rename `retiredIn` to `retirementEpoch` in `PoolRetirementCertificate`.

The phrase "retired in" is somewhat ambiguous, as it might lead readers
to believe that a pool has already retired, when in fact this value can
refer to an epoch that is still in the future.

In comparison, the phrase "retirement epoch" is neutral w.r.t. whether
the epoch is in the future or the past.

rvl commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -421,6 +419,20 @@ newDBLayer trace fp timeInterpreter = do
+                    [ "SELECT * FROM "
+                    , databaseViewName activePoolRetirements
+                    , "WHERE retirement_epoch <="
+                    , T.toText epochNo
```
Isn't there a `?` parameter syntax that we should be using here.

jonathanknowles replied:
> Isn't there a `?` parameter syntax that we should be using here.

I don't know, is there?

rvl replied:
Yes we must maintain a good habit of always using `?` value placeholders. Even though in this particular instance it is safe.

Injection is the number 1 OWASP risk.

Also I am not fond of using views when normal queries would work just as well.

jonathanknowles replied:
> Yes we must maintain a good habit of always using ? value placeholders. Even though in this particular instance it is safe.

Fair enough.

Since this PR no longer touches this area of the code, I will create a separate PR to add `?` value placeholders.

**EDIT: See separate PR here: https://github.com/input-output-hk/cardano-wallet/pull/2053**

rvl commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -383,16 +370,27 @@ newDBLayer trace fp timeInterpreter = do
+                    report "Found no retired pools."
+                    pure []
+                retirementCerts -> do
+                    report $ T.unlines
```
Part of the benefit of typed tracing is that you don't need to obscure the code with log string formatting.

I would make two messages:
 - `MsgRemoveRetiredPools EpochNumber BracketLog` (covers start and finish - use with `bracketTracer`)
 - `MsgRemoveRetiredPoolsCerts [Cert]`

Then put all the message formatting in the ToText instance.

jonathanknowles replied:
@rvl wrote:
> Part of the benefit of typed tracing is that you don't need to obscure the code with log string formatting.

Thanks for this feedback.

I do agree with you. However, one issue is that `DBLog` type is _general_ to all databases. Ideally, `DBLog` itself shouldn't use types that are specific to _pools_ or _wallets_. 

An obvious solution would be to introduce `WalletDbLog` and `PoolDbLog` types, of which the latter would have the constructor that you suggest: `MsgRemoveRetiredPools`. However, this would require changing references to `DBLog` in quite a few places. If we decide to do this, I suggest we create a **separate PR** (or make an item of technical debt), in order to avoid complicating this PR too much.

> I would make two messages:
>   * `MsgRemoveRetiredPools EpochNumber BracketLog` (covers start and finish - use with `bracketTracer`)
>   * `MsgRemoveRetiredPoolsCerts [Cert]`

Thanks for this suggestion: I didn't know about `bracketTracer`.

I've adjusted `removeRetiredPools` to use bracket-style tracing, with the following general constructors (defined within `DBLog`):

```hs
| MsgGarbageCollection Text BracketLog
| MsgGarbageCollectionStep Text
```

See: c912a269501343563283281b5771c66b619e5d32

jonathanknowles replied:
I've created a technical debt ticket to cover the idea of defining separate `WalletDbLog` and `PoolDbLog` types here: https://jira.iohk.io/browse/ADP-407.

While tackling the above ticket, we could also go through the existing `DbLog` constructors and replace the "stringly-typed" constructors with strongly-typed equivalents in `WalletDbLog` and `PoolDbLog`. Hopefully, this should also clean up parts of the code that are obscured with log formatting logic.

jonathanknowles replied:
@rvl we now have the following types:

https://github.com/input-output-hk/cardano-wallet/blob/ab68664b81304faaa3037ce5c1e7a772f555ba45/lib/core/src/Cardano/Pool/DB/Log.hs#L35-L36

jonathanknowles committed:
Make `removeRetiredPools` backend-agnostic.

jonathanknowles committed:
Introduce `MsgGarbageCollectionStep`.

In preparation for bracket-style logging.

jonathanknowles committed:
Use bracket-style logging in `removeRetiredPools`.

In response to review feedback here:

https://github.com/input-output-hk/cardano-wallet/pull/2047#discussion_r471933288

jonathanknowles committed:
Add `MsgRemovingRetiredPools` constructor to `PoolDbLog` type.

jonathanknowles committed:
Move `PoolDbLog` to separate module.

To avoid cyclic module dependencies.

jonathanknowles committed:
Use `PoolDbLog` in `removeRetiredPools`.

jonathanknowles committed:
Use `[PoolRetirementCertificate]` in `MsgRemovingRetiredPools`.

jonathanknowles committed:
Use `MsgRemovingRetiredPools` in `removeRetiredPools`.

jonathanknowles committed:
Add constructor `MsgRemovingRetiredPoolsForEpoch` to `PoolDbLog`.

jonathanknowles committed:
Use `MsgRemovingRetiredPoolsForEpoch` in `removeRetiredPools`.

jonathanknowles committed:
Removing unused `MsgGarbageCollection[Step]` constructors.

jonathanknowles committed:
Remove unused LANGUAGE pragma.

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -280,6 +288,33 @@ determinePoolLifeCycleStatus mReg mRet = case (mReg, mRet) of
+    action = atomically (listRetiredPools epoch) >>=
+        \retirementCerts -> do
+            traceWith trace $ MsgRemovingRetiredPools retirementCerts
+            atomically $ removePools (view #poolId <$> retirementCerts)
```
The list and remove queries should be within a single `atomically` block. You should be able to liftIO the logging statement.

jonathanknowles replied:
> The list and remove queries should be within a single atomically block. You should be able to liftIO the logging statement.

Good point.

Fixed in 5f9c7df9f9a7c66c38d7bc74ad502737ec12febc.

jonathanknowles committed:
List pools and remove them in a single atomic operation.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2047#discussion_r475426914

rvl commented on lib/core/src/Cardano/Pool/DB/Log.hs:
```diff
@@ -0,0 +1,67 @@
+        MsgGeneric e -> getSeverityAnnotation e
+        MsgRemovingPool {} -> Notice
+        MsgRemovingRetiredPools {} -> Notice
+        MsgRemovingRetiredPoolsForEpoch {} -> Notice
```
I don't think all of these should be `Notice` level. Ideally there should be one notice for when a pool is registered and just one more notice for when it retires. The rest is probably debug.

jonathanknowles replied:
> I don't think all of these should be Notice level. Ideally there should be one notice for when a pool is registered and just one more notice for when it retires. The rest is probably debug.

Fixed in 024d36b7c51dac6cb9624697c88fcfa32b82f997.

rvl commented on lib/core/src/Cardano/Pool/DB/Log.hs:
```diff
@@ -0,0 +1,67 @@
+instance ToText PoolDbLog where
+    toText = \case
+        MsgGeneric e -> toText e
+        MsgRemovingPool p -> mconcat
+            [ "Removing the following pool from the database: "
+            , toText p
+            , "."
+            ]
+        MsgRemovingRetiredPools [] ->
+            "There are no retired pools to remove."
+        MsgRemovingRetiredPools poolRetirementCerts -> T.unlines
+            [ "Removing the following retired pools:"
+            , T.unlines (pretty <$> poolRetirementCerts)
+            ]
+        MsgRemovingRetiredPoolsForEpoch epoch nestedMessage -> T.concat
+            [ "Removing pools that retired in or before epoch "
+            , toText epoch
+            , ": "
+            , toText nestedMessage
+            ]
```
The [`Fmt`](http://hackage.haskell.org/package/fmt-0.6.1.2/docs/Fmt.html) module has better functions for constructing string messages.

jonathanknowles replied:
> The [`Fmt`](http://hackage.haskell.org/package/fmt-0.6.1.2/docs/Fmt.html) module has better functions for constructing string messages.

Thanks. I've made a note to research this a bit more.

rvl reviewed (approved):
Two comments about logging for you to address. Since `removeRetiredPools` is not used anywhere I assume there will be another PR.

jonathanknowles committed:
Reduce log `Severity` to `Debug` for pool retirement messages.

For:

- MsgRemovingRetiredPools
- MsgRemovingRetiredPoolsForEpoch

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2047#discussion_r475486039
